cluster-autoscaler/processors/datadog/pods/filter_long_pending.go

```diff
@@ -58,45 +79,98 @@ func (p *filterOutLongPending) CleanUp() {}
-				lastTried: now(),
+		refID := pod.GetUID()
+		controllerRef := metav1.GetControllerOf(pod)
+		if controllerRef != nil {
+			refID = controllerRef.UID
+		}
+
+		// ephemal ownerrefs -> pods mapping
+		if _, ok := refsToPods[refID]; !ok {
+			refsToPods[refID] = make([]*apiv1.Pod, 0)
+		}
+		refsToPods[refID] = append(refsToPods[refID], pod)
+
+		// track upscale attempts for a given ownerref
+		if _, ok := p.seen[refID]; !ok {
+			p.seen[refID] = &pendingTracker{
+				lastSeen: now(),
 			}
 		}
 
-		if p.seen[pod.UID].firstSeen.Add(longPendingCutoff).Before(now()) {
-			deadline := p.seen[pod.UID].lastTried.Add(longPendingBackoff)
-			if deadline.After(now()) {
-				klog.Warningf("ignoring long pending pod %s/%s until %s",
-					pod.GetNamespace(), pod.GetName(), deadline)
-				continue
+		// reset counters whenever a workload has new pods
+		if p.seen[refID].lastSeen.Before(pod.GetCreationTimestamp().Time) {
+			p.seen[refID] = &pendingTracker{
+				lastSeen: pod.GetCreationTimestamp().Time,
 			}
 		}
-		p.seen[pod.UID].lastTried = now()
+	}
+
+	for ref, pods := range refsToPods {
+		// skip (don't attempt upscales for) quarantined pods
+		if p.seen[ref].nextTry.After(now()) {
+			logSkipped(ref, p.seen[ref], pods)
+			continue
+		}
 
-		allowedPods = append(allowedPods, pod)
+		// let autoscaler attempt upscales for non-quarantined pods
+		allowedPods = append(allowedPods, pods...)
+		p.seen[ref].attempts++
+
+		// don't quarantine pods until they had several upscale opportunities
+		if p.seen[ref].attempts < minAttempts {
+			continue
+		}
+
+		// don't quarantine workloads having recent pods
+		if p.seen[ref].lastSeen.Add(longPendingCutoff).After(now()) {
+			continue
+		}
+
+		// pending pods from that workload already had several upscale attempts
+		// opportunities, and arent't recent: will delay after current loop.
+		p.seen[ref].nextTry = p.deadlineFunc(p.seen[ref].attempts, coolDownDelay)
 	}
 
-	for uid := range p.seen {
-		if _, found := currentPods[uid]; !found {
-			delete(p.seen, uid)
+	// reap stale entries
+	for ref := range p.seen {
+		if _, ok := refsToPods[ref]; !ok {
+			delete(p.seen, ref)
 		}
 	}
 
 	return allowedPods, nil
 }
+
+func logSkipped(ref types.UID, tracker *pendingTracker, pods []*apiv1.Pod) {
+	klog.Warningf(
+		"ignoring %d long pending pod(s) (eg. %s/%s) "+
+			"from workload uid %q until %q (attempted %d times)",
+		len(pods), pods[0].GetNamespace(), pods[0].GetName(),
+		ref, tracker.nextTry, tracker.attempts,
+	)
+}
+
+func buildDeadline(attempts int, coolDownDelay time.Duration) time.Time {
+	increment := time.Duration(attempts-minAttempts) * retriesIncrement
+	delay := minDuration(increment, 2*coolDownDelay.Abs()) - jitterDuration(coolDownDelay.Abs())
+	return now().Add(delay.Abs())
+}
+
+func minDuration(a, b time.Duration) time.Duration {
+	if a <= b {
+		return a
+	}
+	return b
+}
+
+func jitterDuration(duration time.Duration) time.Duration {
+	rand.Seed(time.Now().UnixNano())
```

Comment:
I think you should only seed the random number generator once at the start of the program.